### ISSUE TITLE
Notes with lines longer than 256 character truncated, declared corrupt, after upgrade to 0.28

### ISSUE THREAD

**Issue by dajoker** [tomboy-notes/tomboy-ng#173]
openSUSE 15.1 x86_64
tomboy-ng 0.28 (upgraded from 0.27), GTK2 x86_64 version

I saw the note about the updated version and decided to upgrade.  I backed up my notes first as I always do, and then upgraded the RPM and restarted Tomboy-ng.  Upon startup I was presented with a screen indicating I could not go on because of corrupt notes.  Going in to view the corrupt notes I found a few things which are disturbing.

1.  The XML for the first one is valid.  The problem appears to be that two of the lines are no longer separated by a newline, but the XML (per xmllint) is still valid.
2.  The joining of those two lines is new, meaning my backed-up copy does not have the two lines joined; its XML is also valid (per xmllint).
3.  I have dozens of files with updated timestamps as of the time I loaded Tomboy-ng 0.28 for the first time, including these two which are marked as corrupt (see screenshot showing install time and timestamp of dozens of files).
4.  The second file marked corrupt has data lost at the point where I had two ampersands (&amp;&amp) and now where I only have &amp (no semi-colon or subsequent data).

Consequently I think this release should be pulled.  I'm reverting to my 0.27 release for now and restoring from my backup.  Sorry.

**Comment by dajoker** [comment]
The files I meant to attach before; the first shows the corrupt line which is not corrupt; the second shows the files modified on the first load along with the install timestamp from the RPM system, and the third shows a screenshot of the errors received when starting Tomboy-ng.

[20200418-231A.txt](https://github.com/tomboy-notes/tomboy-ng/files/4498395/20200418-231A.txt)
[20200418-file-mods-on-first-load.txt](https://github.com/tomboy-notes/tomboy-ng/files/4498396/20200418-file-mods-on-first-load.txt)
![20200418-ss0](https://user-images.githubusercontent.com/307648/79680036-6966ab80-81c8-11ea-8581-4889af370f6e.png)

**Comment by dajoker** [comment]
I comparing my backup with my originals and lines longer than 256 character (e.g. the expected semicolon from the screenshot above) are truncated.

**Comment by davidbannon** [comment]
oh, thats just about as bad as it gets.
I have some code in there that corrects a bad date stamp that I have seen in a number of notes and was reported by another user.  Needless to say, I tested and tested and saw no problems and really don't expect anyone else to have that bad date stamp anyway (it came form a bug in an early release).
Could you send me one of the notes it appears to be changing ?  Ideally both the unchanged one and the changed one. But they will have the same file name so, just the unchanged one will do. Without that, I am afraid its going to be very hard to replicate the problem.
From memory, you use windows ?  You should be able to grab one note and attach it to an email, send direct to me rather than to this issues list for privacy.
David

**Comment by davidbannon** [comment]
I have put big warning on release pages, if I withdraw that release, people will just get previous one thing its new.

**Comment by davidbannon** [comment]
Looks to me as if the monospace tag has had a couple of spaces inserted into it.  But that may be an artifact of how you pasted it in perhaps ?   I see -
.... mon    ospace....

 and thats going to generate bad xml.  I really need to see one of your notes that has not been changed but will be if v0.28 has its evil way with it.   Does not look like its associated with that date fixing thing I mentioned.

**Comment by davidbannon** [comment]
OK, I got it.  
Yes, it was the code correcting an unimportant mistake in the date stamp.
Problem related to the use of a traditional 255 character Pascal string rather the the unlimited length string we have been using by default for, perhaps 20 years ??
I am so sorry and incredibly grateful you picked this up so quickly !  (hmm, and it means you had some notes with that minor error in their date...)
Watch for version 0.28a in an hour or so.
Davo

**Comment by dajoker** [comment]
0.28a seems to be better, though I must admit I wish the reformatting of these 105 notes (of mine) would have at least resulted in the same timestamp value since the only change was metadata.  Oh well; it's done now and I do not expect you'll want to fix it again adding in parsing of old dates just to preserve them at this point.  I'll do some more comparisons and testing, but at least so far I do not see the 255 character truncation, or the resulting corruption errors, that I noticed before.  Thanks for the quick fix.

I'm not closing just since I do not want to interfere with your workflow, but I consider this bug closed now.

**Comment by davidbannon** [comment]
yes, I could have done a lot better with this particular problem. In hindsight what I should have done was leave the existing date, hours and minutes and then just added the necessary milliseconds and timezone information based on 'now'. ....

**Comment by davidbannon** [comment]
OK, ver 0.28b is a bit more sensitive about the historical data.
It will retain date, hours and minutes of timestamp but make it look like it should be.  I have updated the linux and Mac packages, pretty sure there was not a Mac version of -ng when the cause of this problem was around.  Will do Mac later today just for consistency.
dajoke, not sure if you want to do yet another restore, upgrade cycle .....

**Comment by dajoker** [comment]
Thanks for taking the time to do that; I appreciate it.

I reverted to the backed-up notes, upgraded Tomboy-ng, and compared again and now the timestamps are back as they were but with the fixed formatting so now the metadata are the same in meaning which is great as my notes are showing correctly in the tray icon again.  A sample of the 'diff' output follows just to make it clear what changed on the files:

`diff --minimal -r tomboy-ng-20200418/0380ed61-784e-4dec-ab52-16ab40a6b9ec.note tomboy-ng/0380ed61-784e-4dec-ab52-16ab40a6b9ec.note`
`333c333`
`<   <last-change-date>2019-12-15 20:33:16</last-change-date>`
`---`
`>   <last-change-date>2019-12-15T20:33:16.3790000-06:00</last-change-date>`
`343c343`
`< </note>`
`\ No newline at end of file`

**Comment by davidbannon** [comment]
Yep, thats correct, thanks for that confirmation. I have added in a 'T' between date and time and I have added milliseconds and timezone at the end.  Obviously the number of milliseconds is wrong but having them there with a correct timezone is important when syncing.
But you have rocked me to the core with the date, that original error was added to your note late 2019 ?   Any chance you were using a very early tomboy-ng about then ?   I was quite sure that this problem was introduced in the very early stages ......

**Comment by dajoker** [comment]
Nope, no chance, but in fact I was using the original Tomboy (mono-based) from my first use whenever many years ago until I upgraded to tomboy-ng with my system upgrade on 2019-12-30 or so.  Here is another diff snippet that may shed more light on this:

`diff --minimal -r tomboy-ng-20200418/f740753b-ebe9-445a-9ed3-1cf4446a7132.note tomboy-ng/f740753b-ebe9-445a-9ed3-1cf4446a7132.note`
`83c83`
`<   <last-change-date>2017-08-27 06:15:49</last-change-date>`
`---`
`>   <last-change-date>2017-08-27T06:15:49.3590000-06:00</last-change-date>`
`93c93`
`< </note>`
`\ No newline at end of file`
`---`
`> </note>`

Now I have looked a bit more at this note above and am surprised that its format changed at some point after my upgrade to tomboy-ng on 2019-12-30.  The metadata from my pre-tomboy-ng backup (taken because I was upgrading my whole system, so I have that backup still) follows:

`  <last-change-date>2017-08-27T06:15:49.1096530-06:00</last-change-date>`
`  <last-metadata-change-date>2017-08-27T06:15:49.1103110-06:00</last-metadata-change-date>`
`  <create-date>2017-08-24T07:35:50.3592290-06:00</create-date>`
`  <cursor-position>7429</cursor-position>`
`  <selection-bound-position>8730</selection-bound-position>`
`  <width>573</width>`
`  <height>813</height>`
`  <x>1306</x>`
`  <y>83</y>`
`  <open-on-startup>False</open-on-startup>`

**Comment by dajoker** [comment]
I've confirmed that all 105 of the notes that were changed had correct timestamps prior to my first use of tomboy-ng on 2019-12-30 or shortly thereafter:

On my backup directory prior to the change the following command returns nothing, but on my backup directory from last week it returns the 105 shown:

> grep 'last-change-date' * | grep -v 'T' | wc -l
> 105

**Comment by davidbannon** [comment]
Wow, love people who keep backups !  
I will now have to trawl through git and see just when that problem code existed.  
Thanks

**Comment by dajoker** [comment]
Just in case you're not familiar with it, 'git blame path/to/object' is just wonderful, showing line-by-line commits, and dates, and stuff.  You can also specify a particular commit and have it show the "blame" output from that point back too:

git blame commit-id-here path/to/object

**Comment by davidbannon** [comment]
OK, well, I really don't understand. The code causing the underlying problem of  the bad dates  had been fixed before I made my first commit to github, November 2017.   I have been back through my early backups and my incidents of this problem are all prior to this date. 
Indeed, two actually predate tomboy-ng itself.  So, I am really stumped !
I have to wonder now if there is something else causing a this problem.  I will try and find it because a bad date us just not on ....
